Bottom: 2f30210c63546db6488c195c7655365cf47fb9a0
Top:    b52fecdb18bcb46d8adba16e1f550b91bf8238da
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2012-11-21 20:26:36 +0000

Move HID drawing API prototypes into a separate header file

Due to the mess of places we define various things in and the order
we pull in headers, we cannot easily create APIs in hid.h which rely
on being passed PCB data-structures.


---

diff --git a/src/crosshair.c b/src/crosshair.c
index 65a94fb..1079c5f 100644
--- a/src/crosshair.c
+++ b/src/crosshair.c
@@ -36,6 +36,7 @@
 #include <math.h>
 
 #include "global.h"
+#include "hid_draw.h"
 
 #include "crosshair.h"
 #include "data.h"
diff --git a/src/draw.c b/src/draw.c
index e4d59f5..32aacd5 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -33,6 +33,7 @@
 #endif
 
 #include "global.h"
+#include "hid_draw.h"
 
 /*#include "clip.h"*/
 #include "compat.h"
diff --git a/src/hid.h b/src/hid.h
index 6135750..003b6a4 100644
--- a/src/hid.h
+++ b/src/hid.h
@@ -232,69 +232,8 @@ typedef enum
     int (*throw_drc_dialog) (void);
   } HID_DRC_GUI;
 
-  enum mask_mode {
-    HID_MASK_OFF    = 0, /* Flush the buffer and return to non-mask operation. */
-    HID_MASK_BEFORE = 1, /* Polygons being drawn before clears.                */
-    HID_MASK_CLEAR  = 2, /* Clearances being drawn.                            */
-    HID_MASK_AFTER  = 3, /* Polygons being drawn after clears.                 */
-  };
-
-/* Low level drawing API */
-  typedef struct
-  {
-    /* Drawing Functions.  Coordinates and distances are ALWAYS in PCB's
-       default coordinates (1 nm at the time this comment was written).
-       Angles are always in degrees, with 0 being "right" (positive X)
-       and 90 being "up" (positive Y).  */
-
-    /* Make an empty graphics context.  */
-    hidGC (*make_gc) (void);
-    void (*destroy_gc) (hidGC gc_);
-    void (*use_mask) (enum mask_mode mode);
-
-    /* Set a color.  Names can be like "red" or "#rrggbb" or special
-       names like "erase".  *Always* use the "erase" color for removing
-       ink (like polygon reliefs or thermals), as you cannot rely on
-       knowing the background color or special needs of the HID.  Always
-       use the "drill" color to draw holes.  You may assume this is
-       cheap enough to call inside the redraw callback, but not cheap
-       enough to call for each item drawn. */
-    void (*set_color) (hidGC gc, const char *name);
-
-    /* Set the line style.  While calling this is cheap, calling it with
-       different values each time may be expensive, so grouping items by
-       line style is helpful.  */
-    void (*set_line_cap) (hidGC gc, EndCapStyle style);
-    void (*set_line_width) (hidGC gc, Coord width);
-    void (*set_draw_xor) (hidGC gc, int xor_);
-
-    /* Blends 20% or so color with 80% background.  Only used for
-       assembly drawings so far. */
-    void (*set_draw_faded) (hidGC gc, int faded);
-
-    /* The usual drawing functions.  "draw" means to use segments of the
-       given width, whereas "fill" means to fill to a zero-width
-       outline.  */
-    void (*draw_line)    (hidGC gc, Coord x1, Coord y1, Coord x2, Coord y2);
-    void (*draw_arc)     (hidGC gc, Coord cx, Coord cy, Coord xradius, Coord yradius, Angle start_angle, Angle delta_angle);
-    void (*draw_rect)    (hidGC gc, Coord x1, Coord y1, Coord x2, Coord y2);
-    void (*fill_circle)  (hidGC gc, Coord cx, Coord cy, Coord radius);
-    void (*fill_polygon) (hidGC gc, int n_coords_, Coord *x, Coord *y);
-    void (*fill_rect)    (hidGC gc, Coord x1, Coord y1, Coord x2, Coord y2);
-
-    /* The following APIs render using PCB data-structures, not immediate parameters */
-
-    void (*fill_pcb_polygon) (hidGC gc, PolygonType *poly, const BoxType *clip_box);
-    void (*thindraw_pcb_polygon) (hidGC gc, PolygonType *poly, const BoxType *clip_box);
-    void (*fill_pcb_pad) (hidGC gc, PadType *pad, bool clip, bool mask);
-    void (*thindraw_pcb_pad) (hidGC gc, PadType *pad, bool clip, bool mask);
-    void (*fill_pcb_pv) (hidGC fg_gc, hidGC bg_gc, PinType *pv, bool drawHole, bool mask);
-    void (*thindraw_pcb_pv) (hidGC fg_gc, hidGC bg_gc, PinType *pv, bool drawHole, bool mask);
-
-  } HID_DRAW_API;
-
-
   typedef struct hid_st HID;
+  typedef struct hid_draw_st HID_DRAW;
 
 /* This is the main HID structure.  */
   struct hid_st
@@ -379,7 +318,7 @@ typedef enum
     void (*end_layer) (void);
 
 
-    HID_DRAW_API *graphics;
+    HID_DRAW *graphics;
 
     /* This is for the printer.  If you call this for the GUI, xval and
        yval are ignored, and a dialog pops up to lead you through the
diff --git a/src/hid/common/extents.c b/src/hid/common/extents.c
index 82d7955..e497573 100644
--- a/src/hid/common/extents.c
+++ b/src/hid/common/extents.c
@@ -11,6 +11,7 @@
 #include "data.h"
 
 #include "hid.h"
+#include "hid_draw.h"
 #include "../hidint.h"
 #include "hid/common/draw_helpers.h"
 
diff --git a/src/hid/common/hidnogui.c b/src/hid/common/hidnogui.c
index 2123ad0..db38f56 100644
--- a/src/hid/common/hidnogui.c
+++ b/src/hid/common/hidnogui.c
@@ -9,6 +9,7 @@
 
 #include "global.h"
 #include "hid.h"
+#include "hid_draw.h"
 
 #ifdef HAVE_LIBDMALLOC
 #include <dmalloc.h>
diff --git a/src/hid/gcode/gcode.c b/src/hid/gcode/gcode.c
index 952621f..f62ceb6 100644
--- a/src/hid/gcode/gcode.c
+++ b/src/hid/gcode/gcode.c
@@ -53,6 +53,7 @@
 #include "rats.h"
 
 #include "hid.h"
+#include "hid_draw.h"
 #include "../hidint.h"
 #include <gd.h>
 #include "hid/common/hidnogui.h"
diff --git a/src/hid/gerber/gerber.c b/src/hid/gerber/gerber.c
index 5f5aebc..66c4ed1 100644
--- a/src/hid/gerber/gerber.c
+++ b/src/hid/gerber/gerber.c
@@ -27,6 +27,7 @@
 #include "pcb-printf.h"
 
 #include "hid.h"
+#include "hid_draw.h"
 #include "../hidint.h"
 #include "hid/common/hidnogui.h"
 #include "hid/common/draw_helpers.h"
diff --git a/src/hid/lesstif/main.c b/src/hid/lesstif/main.c
index f1b8308..5b0fdce 100644
--- a/src/hid/lesstif/main.c
+++ b/src/hid/lesstif/main.c
@@ -25,6 +25,7 @@
 #include "error.h"
 
 #include "hid.h"
+#include "hid_draw.h"
 #include "../hidint.h"
 #include "hid/common/hidnogui.h"
 #include "hid/common/draw_helpers.h"
diff --git a/src/hid/nelma/nelma.c b/src/hid/nelma/nelma.c
index 6c837be..31aa902 100644
--- a/src/hid/nelma/nelma.c
+++ b/src/hid/nelma/nelma.c
@@ -70,6 +70,7 @@
 #include "rats.h"
 
 #include "hid.h"
+#include "hid_draw.h"
 #include "../hidint.h"
 #include "hid/common/hidnogui.h"
 #include "hid/common/draw_helpers.h"
diff --git a/src/hid/png/png.c b/src/hid/png/png.c
index 62cb62f..d333495 100644
--- a/src/hid/png/png.c
+++ b/src/hid/png/png.c
@@ -41,6 +41,7 @@
 #include "misc.h"
 
 #include "hid.h"
+#include "hid_draw.h"
 #include "../hidint.h"
 #include "hid/common/hidnogui.h"
 #include "hid/common/draw_helpers.h"
diff --git a/src/hid/ps/eps.c b/src/hid/ps/eps.c
index 1a43106..b5994c6 100644
--- a/src/hid/ps/eps.c
+++ b/src/hid/ps/eps.c
@@ -14,6 +14,7 @@
 #include "pcb-printf.h"
 
 #include "hid.h"
+#include "hid_draw.h"
 #include "../hidint.h"
 #include "hid/common/hidnogui.h"
 #include "hid/common/draw_helpers.h"
diff --git a/src/hid/ps/ps.c b/src/hid/ps/ps.c
index 6930507..7015ffd 100644
--- a/src/hid/ps/ps.c
+++ b/src/hid/ps/ps.c
@@ -17,6 +17,7 @@
 #include "pcb-printf.h"
 
 #include "hid.h"
+#include "hid_draw.h"
 #include "../hidint.h"
 #include "hid/common/hidnogui.h"
 #include "hid/common/draw_helpers.h"
diff --git a/src/print.c b/src/print.c
index 27e1c98..fc6f9ff 100644
--- a/src/print.c
+++ b/src/print.c
@@ -38,6 +38,7 @@
 #endif
 
 #include "global.h"
+#include "hid_draw.h"
 
 #include <time.h>
 #ifdef HAVE_UNISTD_H
